letsencrypt/client/tests/standalone_authenticator_test.py

```diff
@@ -0,0 +1,564 @@
+        cert_msg = tls_generate_cert_msg(cert)
+        self.assertEqual(cert_msg.encode("hex"),
+                         "16030301ec0b0001e80001e50001e2308201de30820188a003"
+                         "02010202020539300d06092a864886f70d01010b0500307731"
+                         "0b30090603550406130255533111300f06035504080c084d69"
+                         "63686967616e3112301006035504070c09416e6e204172626f"
+                         "72312b3029060355040a0c22556e6976657273697479206f66"
+                         "204d6963686967616e20616e64207468652045464631143012"
+                         "06035504030c0b6578616d706c652e636f6d301e170d313431"
+                         "3231313232333434355a170d3134313231383232333434355a"
+                         "3077310b30090603550406130255533111300f06035504080c"
+                         "084d6963686967616e3112301006035504070c09416e6e2041"
+                         "72626f72312b3029060355040a0c22556e6976657273697479"
+                         "206f66204d6963686967616e20616e64207468652045464631"
+                         "14301206035504030c0b6578616d706c652e636f6d305c300d"
+                         "06092a864886f70d0101010500034b003048024100ac7573b4"
+                         "51ed1fddae705243fcdfc75bd02c751b14b875010410e51f03"
+                         "6545dddfa79f34aefdbee90584df471681d9894bce8e6d1cfa"
+                         "9544e8af84744fedc2e50203010001300d06092a864886f70d"
+                         "01010b05000341002db8cf421dc0854a4a59ed92c965bebeb3"
+                         "25ea411f97cc9dd7e4dd7269d748d3e9513ed7828db63874d9"
+                         "ae7a1a8ada02f2404f9fc7ebb13c1af27fa1c36707fa")
+
+
+class TLSServerHelloDoneTest(unittest.TestCase):
+    # pylint: disable=too-few-public-methods
+    """Tests for tls_generate_server_hello_done() function."""
+    def test_tls_generate_server_hello_done(self):
+        from letsencrypt.client.standalone_authenticator import \
+            tls_generate_server_hello_done
+        self.assertEqual(tls_generate_server_hello_done().encode("hex"), \
+            "16030300040e000000")
+
+
+class ChallPrefTest(unittest.TestCase):
+    """Tests for chall_pref() method."""
+    def setUp(self):
+        from letsencrypt.client.standalone_authenticator import \
+            StandaloneAuthenticator
+        self.authenticator = StandaloneAuthenticator()
+
+    def test_chall_pref(self):
+        self.assertEqual(self.authenticator.get_chall_pref("example.com"),
+                         ["dvsni"])
+
+
+class SNICallbackTest(unittest.TestCase):
+    """Tests for sni_callback() method."""
+    def setUp(self):
+        from letsencrypt.client.standalone_authenticator import \
+            StandaloneAuthenticator
+        from letsencrypt.client.challenge_util import dvsni_gen_cert
+        from letsencrypt.client import le_util
+        import OpenSSL.crypto
+        from OpenSSL.crypto import FILETYPE_PEM
```

Comment:
modules that are not under test (here: `OpenSSL`, `le_util`, `challenge_util`) should be imported at the top